File: fastapi-backend/app/services/addon_service.py
"""
Addon Service

Business logic for addon operations
"""

from typing import List, Optional
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from app.models.addon import Addon
from app.schemas.addon import AddonCreate, AddonUpdate


class AddonService:
    """Service for addon operations"""

    @staticmethod
    async def get_all_addons(
        db: AsyncSession,
        available_only: bool = True
    ) -> List[Addon]:
        """
        Get all addons

        Args:
            db: Database session
            available_only: Only return available addons

        Returns:
            List of Addon objects
        """
        query = select(Addon)

        if available_only:
            query = query.where(Addon.available == True)

        query = query.order_by(Addon.category, Addon.name)

        result = await db.execute(query)
        return list(result.scalars().all())

    @staticmethod
    async def get_addon_by_id(db: AsyncSession, addon_id: int) -> Optional[Addon]:
        """Get a single addon by ID"""
        query = select(Addon).where(Addon.id == addon_id)
        result = await db.execute(query)
        return result.scalar_one_or_none()

    @staticmethod
    async def create_addon(db: AsyncSession, addon_data: AddonCreate) -> Addon:
        """
        Create a new addon

        Args:
            db: Database session
            addon_data: Addon creation data (price is in pesos)

        Returns:
            Created Addon object
        """
        # Convert price from pesos to cents for storage
        price_in_cents = int(addon_data.price * 100)

        new_addon = Addon(
            name=addon_data.name,
            description=addon_data.description,
            price=price_in_cents,
            category=addon_data.category,
            available=addon_data.available,
            max_quantity=addon_data.max_quantity
        )

        db.add(new_addon)
        await db.commit()
        await db.refresh(new_addon)

        return new_addon

    @staticmethod
    async def update_addon(
        db: AsyncSession,
        addon_id: int,
        addon_data: AddonUpdate
    ) -> Optional[Addon]:
        """
        Update an existing addon

        Args:
            db: Database session
            addon_id: Addon ID
            addon_data: Addon update data (price is in pesos)

        Returns:
            Updated Addon object or None if not found
        """
        addon = await AddonService.get_addon_by_id(db, addon_id)

        if not addon:
            return None

        # Update fields
        if addon_data.name is not None:
            addon.name = addon_data.name
        if addon_data.description is not None:
            addon.description = addon_data.description
        if addon_data.price is not None:
            addon.price = int(addon_data.price * 100)  # Convert to cents
        if addon_data.category is not None:
            addon.category = addon_data.category
        if addon_data.available is not None:
            addon.available = addon_data.available
        if addon_data.max_quantity is not None:
            addon.max_quantity = addon_data.max_quantity

        await db.commit()
        await db.refresh(addon)

        return addon

    @staticmethod
    async def delete_addon(db: AsyncSession, addon_id: int) -> bool:
        """
        Delete an addon

        Args:
            db: Database session
            addon_id: Addon ID

        Returns:
            True if deleted, False if not found
        """
        addon = await AddonService.get_addon_by_id(db, addon_id)

        if not addon:
            return False

        await db.delete(addon)
        await db.commit()

        return True


# Singleton instance
addon_service = AddonService()

File: quickserve-frontend/lib/api/services/addonService.ts
/**
 * Addon Service
 *
 * Handles API calls for addon management
 */

import { apiClient } from '../client'

export interface Addon {
  id: number
  name: string
  description: string | null
  price: number
  price_in_pesos: number
  category: string
  available: boolean
  max_quantity: number
  created_at: string
  updated_at: string
}

export interface AddonCreate {
  name: string
  description?: string
  price: number
  category?: string
  available?: boolean
  max_quantity?: number
}

export interface AddonUpdate {
  name?: string
  description?: string
  price?: number
  category?: string
  available?: boolean
  max_quantity?: number
}

export interface AddonListResponse {
  addons: Addon[]
  total: number
}

/**
 * Addon Service Class
 */
class AddonService {
  private readonly basePath = '/addons'

  /**
   * Get all addons
   */
  async getAddons(availableOnly: boolean = true): Promise<AddonListResponse> {
    const params = new URLSearchParams()
    if (availableOnly) {
      params.append('available_only', 'true')
    }

    const queryString = params.toString()
    const response = await apiClient.get<AddonListResponse>(
      `${this.basePath}${queryString ? `?${queryString}` : ''}`
    )
    return response
  }

  /**
   * Get a single addon by ID
   */
  async getAddon(id: number): Promise<Addon> {
    const response = await apiClient.get<Addon>(`${this.basePath}/${id}`)
    return response
  }

  /**
   * Create a new addon
   */
  async createAddon(data: AddonCreate): Promise<Addon> {
    const response = await apiClient.post<Addon>(this.basePath, data)
    return response
  }

  /**
   * Update an existing addon
   */
  async updateAddon(id: number, data: AddonUpdate): Promise<Addon> {
    const response = await apiClient.put<Addon>(`${this.basePath}/${id}`, data)
    return response
  }

  /**
   * Delete an addon
   */
  async deleteAddon(id: number): Promise<void> {
    await apiClient.delete(`${this.basePath}/${id}`)
  }
}

// Export singleton instance
export const addonService = new AddonService()
